lib/tail-file.js

```diff
@@ -0,0 +1,279 @@
+    this[kMaxPollFailures] = maxPollFailures || 10
+    this[kPollFailureCount] = 0
+    this[kStartPos] = null
+    this[kStream] = null
+    this[kFileHandle] = null
+    this[kPollTimer] = null
+    this[kQuitting] = false
+    this[kInode] = null
+  }
+
+  async start() {
+    try {
+      await this._openFile()
+      await this._pollFileForChanges()
+    } catch (err) {
+      return this.quit(err)
+    }
+    return
+  }
+
+  async _openFile() {
+    this[kFileHandle] = await fs.promises.open(this[kFileName])
+    return
+  }
+
+  async _readRemainderFromFileHandle() {
+    // Read the end of a renamed file before re-opening the new file.
+    // Use the file handle since it remains open even if the file name has changed
+    const stats = await this[kFileHandle].stat()
+    const lengthToEnd = stats.size - this[kStartPos]
+    const {buffer} = await this[kFileHandle].read(
+      Buffer.alloc(lengthToEnd)
+    , 0
+    , lengthToEnd
+    , this[kStartPos]
+    )
+    this.push(buffer)
+    await this[kFileHandle].close()
+    this[kFileHandle] = null
+    return
+  }
+
+  async _readChunks(stream) {
+    for await (const chunk of stream) {
+      this[kStartPos] += chunk.length
+      if (!this.push(chunk)) {
+        this[kStream] = stream
+        return
+      }
+    }
+    this[kStream] = null
+    setImmediate(this.emit.bind(this), 'flush')
+    // Immediately try another read
+    this._scheduleTimer(0)
+    return
+  }
+
+  async _pollFileForChanges() {
+    try {
+      const stats = await fs.promises.stat(this[kFileName])
+
+      this[kPollFailureCount] = 0 // reset
+      const eof = stats.size
+      let fileHasChanged = false
+
+      if (this[kStartPos] === null) {
+        // First iteration - nothing has been polled yet
+        this[kStartPos] = eof
+        this[kInode] = stats.ino
+      } else if (this[kInode] !== stats.ino) {
+        // File renamed/rolled
+        await this._readRemainderFromFileHandle()
+        await this._openFile()
+        this[kStartPos] = 0
+        this[kInode] = stats.ino
+        fileHasChanged = true
+        this.emit('renamed', {
+          message: 'The file was renamed or rolled.  Tailing resumed from the beginning.'
+        , filename: this[kFileName]
+        , when: new Date()
+        })
+      } else if (eof < this[kStartPos]) {
+        // Same file, but smaller/truncated
+        this[kStartPos] = 0
+        this[kInode] = stats.ino
+        fileHasChanged = true
+        this.emit('truncated', {
+          message: 'The file was truncated.  Tailing resumed from the beginning.'
+        , filename: this[kFileName]
+        , when: new Date()
+        })
+      } else if (this[kStartPos] !== eof) {
+        fileHasChanged = true
+      }
+
+      if (fileHasChanged) {
+        await this._streamFileChanges()
+      } else {
+        setImmediate(this.emit.bind(this), 'flush')
+      }
+
+      this._scheduleTimer(this[kPollFileIntervalMs])
```

Comment:
@evanlucas you got me thinking about races.  This could be a double timer.  it's scheduled after all the chunks are read, however, this would re-schedule it.  This line should also go inside the else block too, which leaves the scheduling up to `_readChunks()` in the event we're dealing with backpressure and reading.

Reply:
Fixed.

Reply:
Backing this change out.  We need to keep polling so that we don't miss any "new" files coming through.  I tried implementing a "stream queue", but it never got more than 1 queue pushed onto it, so I think the solution here is just to remove the immediate poll schedule after the chunks are read because it's just overwritten as soon as it returns, so that does no good.